src/directives/repeat.ts

```diff
@@ -41,67 +96,139 @@ export function repeat<T>(
-    if (currentMarker !== part.endNode) {
-      removeNodes(container, currentMarker, part.endNode);
-      keyMap.forEach(cleanMap);
+    // Head and tail pointers to old parts and new results
+    let oldStartIndex = 0;
+    let oldStartPart = oldParts[0];
+    let oldEndIndex = oldParts.length - 1;
+    let oldEndPart = oldParts[oldEndIndex];
+    let newStartIndex = 0;
+    let newStartResult = newResults[0];
+    let newEndIndex = newResults.length - 1;
+    let newEndResult = newResults[newEndIndex];
+
+    // Overview of O(n) reconciliation algorithm (general approach based on
+    // ideas found in ivi, vue, snabdom, etc.):
+    // * Iterate old & new lists from both sides, updating, swapping, or
+    // removing
+    //   items at the head/tail locations until neither head nor tail can move
+    // * Once head and tail cannot move, any mismatches are due to either new or
+    //   moved items; create and insert new items, or else move old part to new
+    //   position using the "old key to old index" map.
+    // * If either the new or old pointers move past each other then all we have
+    //   left is additions (if old list exhsusted) or removals (if new list
+    //   exhausted). Those are handled in the final while loops at the end.
+    // * TODO(kschaaf) Note, we could calculate the longest increasing
+    //   subsequence (LIS) of old items in new position, and only move those not
+    //   in the LIS set. However that costs O(nlogn) time and adds a bit more
+    //   code, and only helps make rare types of mutations require fewer moves.
+    //   The above handles removes, adds, reversal, swaps, and single moves of
+    //   contiguous items in linear time, in the minimum number of moves. As
+    //   the number of multiple moves where LIS might help approaches a random
+    //   shuffle, the LIS optimization becomes less helpful, so it seems not
+    //   worth the code at this point, but could consider if a compelling case
+    //   arises.
+    while (oldStartIndex <= oldEndIndex && newStartIndex <= newEndIndex) {
+      if (oldStartPart == null) {
+        // Old part at head has already been used; skip
+        oldStartPart = oldParts[++oldStartIndex];
+      } else if (oldEndPart == null) {
+        // Old part at tail has already been used; skip
+        oldEndPart = oldParts[--oldEndIndex];
+      } else if (oldStartPart.key == newStartResult.key) {
+        // Old head matches new head; update in place
+        newParts[newStartIndex] = updatePart(oldStartPart, newStartResult);
+        oldStartPart = oldParts[++oldStartIndex];
+        newStartResult = newResults[++newStartIndex];
+      } else if (oldEndPart.key == newEndResult.key) {
+        // Old tail matches new tail; update in place
+        newParts[newEndIndex] = updatePart(oldEndPart, newEndResult);
+        oldEndPart = oldParts[--oldEndIndex];
+        newEndResult = newResults[--newEndIndex];
+      } else if (!newKeyToIndexMap.has(oldStartPart.key)) {
```

Comment:
It probably makes sense for removal detection to come first, since any checks against a removed old key that currently come first are repetitive... will move up I guess, then updates will be grouped after that.

Reply:
It'd be great to comment on whether, or which, cases in this if/else chain are order dependant.

Reply:
👍Done.